From 79fd7ddc704daf97929cb7f8ee62a6f89e1805c2 Mon Sep 17 00:00:00 2001
From: Oliver Steffen <osteffen@redhat.com>
Date: Thu, 7 Nov 2024 11:36:22 +0100
Subject: [PATCH] OvmfPkg/ArmVirtPkg: Add a Fallback RNG (RH only)

RH-Author: Oliver Steffen <osteffen@redhat.com>
RH-MergeRequest: 82: Add a Fallback RNG (RH only)
RH-Jira: RHEL-66234
RH-Acked-by: Gerd Hoffmann <kraxel@redhat.com>
RH-Commit: [2/2] ae2c04680e6420e096c667a22c52ec6f6fb46935 (osteffen/edk2)

Since the pixiefail CVE fix, the network stack requires a random number
generator.
In case there is no hardware random number generator available,
have the Platform Boot Manager install a pseudo RNG to ensure
the network can be used.

This patch adds the fallback rng which was introduced in a
previous commit also to the ArmVirtPkg PlatformBootManagerLib.

Signed-off-by: Oliver Steffen <osteffen@redhat.com>

patch_name: edk2-OvmfPkg-ArmVirtPkg-Add-a-Fallback-RNG-RH-only.patch
present_in_specfile: true
location_in_specfile: 49
---
 OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBm.c    | 6 ++++++
 .../PlatformBootManagerLibLight/PlatformBootManagerLib.inf  | 5 +++++
 2 files changed, 11 insertions(+)

diff --git a/OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBm.c b/OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBm.c
index 2c24c65489..273e6f6a7e 100644
--- a/OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBm.c
+++ b/OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBm.c
@@ -30,6 +30,7 @@
 #include <Guid/RootBridgesConnectedEventGroup.h>
 #include <Guid/SerialPortLibVendor.h>
 #include <Library/PlatformBootManagerCommonLib.h>
+#include "FallbackRng.h"
 
 #include "PlatformBm.h"
 
@@ -819,6 +820,7 @@ PlatformBootManagerBeforeConsole (
   //
   FilterAndProcess (&gEfiGraphicsOutputProtocolGuid, NULL, AddOutput);
 
+
   //
   // Add the hardcoded short-form USB keyboard device path to ConIn.
   //
@@ -916,6 +918,8 @@ PlatformBootManagerBeforeConsole (
   //
   FilterAndProcess (&gVirtioDeviceProtocolGuid, IsVirtioSerial, SetupVirtioSerial);
   FilterAndProcess (&gEfiPciIoProtocolGuid, IsVirtioPciSerial, SetupVirtioSerial);
+
+  FallbackRngCheckAndInstall ();
 }
 
 /**
@@ -982,6 +986,8 @@ PlatformBootManagerAfterConsole (
   BOOLEAN        Uninstall;
   BOOLEAN        ShellEnabled;
 
+  FallbackRngPrintWarning ();
+
   //
   // Show the splash screen.
   //
diff --git a/OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBootManagerLib.inf b/OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBootManagerLib.inf
index 6c5552da16..1de19eb507 100644
--- a/OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBootManagerLib.inf
+++ b/OvmfPkg/Library/PlatformBootManagerLibLight/PlatformBootManagerLib.inf
@@ -27,6 +27,8 @@
   PlatformBm.c
   PlatformBm.h
   QemuKernel.c
+  ../PlatformBootManagerLib/FallbackRng.h
+  ../PlatformBootManagerLib/FallbackRng.c
 
 [Packages]
   MdeModulePkg/MdeModulePkg.dec
@@ -54,6 +56,7 @@
   UefiLib
   UefiRuntimeServicesTableLib
   PlatformBootManagerCommonLib
+  RngLib
 
 [FixedPcd]
   gEfiMdePkgTokenSpaceGuid.PcdUartDefaultBaudRate
@@ -72,6 +75,7 @@
   gRootBridgesConnectedEventGroupGuid
   gUefiShellFileGuid
   gUiAppFileGuid
+  gEfiRngAlgorithmRaw
 
 [Protocols]
   gEfiFirmwareVolume2ProtocolGuid
@@ -79,3 +83,4 @@
   gEfiMemoryAttributeProtocolGuid
   gEfiPciRootBridgeIoProtocolGuid
   gVirtioDeviceProtocolGuid
+  gEfiRngProtocolGuid
